xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,305 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+var errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+
+/*
+func (bb) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &gracefulSwitchBalancer{ // Will need to just construct this inline by user/tests
+		cc:              cc,
+		bOpts:           opts,
+		scToSubBalancer: make(map[balancer.SubConn]balancer.Balancer),
+		closed:          grpcsync.NewEvent(),
+	}
+}
+*/
+
+type gracefulSwitchBalancer struct {
+	bOpts          balancer.BuildOptions
+	cc balancer.ClientConn
+
+	balancerCurrent balancer.Balancer
+	balancerPending balancer.Balancer
+	// One mutex...protecting incoming (updateState happening atomically,
+	// scToSubBalancer reads/writes, and also reads from
+	// balancerCurrent/Pending...which will put into a local variable to call
+	// downward into, thus making it so that no lock is held when call downward).
+	mu sync.Mutex
+	scToSubBalancer map[balancer.SubConn]balancer.Balancer
+	pendingState balancer.State
+	currentLbIsReady bool
+
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	if gsb.closed.HasFired() {
+		return
+	}
+
+	// I want this to not intersplice with other updateState() calls and for
+	// this to happen atomically (i.e. if this update isn't atomic writing to
+	// pending state concurrently and sending wrong one out for one
+	// updateState() call).
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+
```

Comment:
In `newSubConn` and `resolverNow`, there are checks to make sure that the call is from either the `current` or the `pending` balancer. And in the former, an error to the same effect is returned as well.

- Is that not check not required here?
- Does it make sense to make a helper function for that check?
- Also, it would be nice if the error message in these cases contains something unique to identify the balancer from which the call came from (maybe a combination of balancer name and pointer).